artofillusion/polymesh/UVMappingCanvas.java

```diff
@@ -272,69 +297,198 @@ private void doPaint(Graphics2D g, boolean export) {
+        for (int i = 0; i < verticesPoints.length; i++) {
+            if (selected[i]) {
+                if (mappingData.meshes[currentPiece].vertices[mappingData.verticesTable[currentPiece][i]].pinned)
+                    g.setColor(pinnedSelectedColor);
+                else
+                    g.setColor(selectedColor);
+
+                g.drawOval(verticesPoints[i].x - 3,
+                        verticesPoints[i].y - 3, 6, 6);
+            } else {
+                if (mappingData.meshes[currentPiece].vertices[mappingData.verticesTable[currentPiece][i]].pinned)
+                    g.setColor(pinnedColor);
+                else
+                    g.setColor(unselectedColor);
+
+                g.fillOval(verticesPoints[i].x - 3,
+                        verticesPoints[i].y - 3, 6, 6);
+            }
+        }
+        if (dragBoxRect != null) {
+            g.setColor(ink);
+            g.drawRect(dragBoxRect.x, dragBoxRect.y, dragBoxRect.width, dragBoxRect.height);
+        }
+        if (manipulator != null)
+            manipulator.paint(g);
+    }
+
+    private void drawGrid(Graphics2D g)
+    {
+        if (! parent.drawGrid())
+            return;
+        if (scale < 2.0)
+            return;
+
+        // Must use the original AT from the Graphics2D because at startup the coordinates
+        // are measured from the content pane of the UVMappingEditorDialog. Later on a new AT would do.
+
+        AffineTransform aBefore = g.getTransform();
+        AffineTransform aNow = new AffineTransform(aBefore);
+        Point corner = VertexToLayout(new Vec2(0, 0));
+        aNow.translate(corner.x, corner.y);
+        aNow.scale(scale, -scale);
+        g.setTransform(aNow);
+        g.setStroke(new BasicStroke((float)(1.0/scale)));
+
+        //double opacy = Math.min(255*Math.sqrt(scale/640), 223);
+        double opacy = Math.min(255.0*Math.pow(scale/520.0, 2.0/3.0), 223.0);
+        Color ink1 = new Color(ink.getRed(), ink.getGreen(), ink.getBlue(), (int)opacy);
+        Color ink2 = new Color(ink.getRed(), ink.getGreen(), ink.getBlue(), (int)(opacy/1.6));
+        Color ink3 = new Color(ink.getRed(), ink.getGreen(), ink.getBlue(), (int)(opacy/2.5));
+        double v = Math.round(vmin -1.0);
+        double u = Math.round(umin -1.0);
+
+        // Avoiding to draw lines over already drawn lines becuse they are transparent
+
+        while (u < umax)
+        {
+            g.setColor(ink1);
+            g.draw(new Line2D.Double(u, vmin, u, vmax));
+            if (scale > 75)
+            {
+                g.setColor(ink2);
+                g.draw(new Line2D.Double(u+0.5, vmin,u+0.5, vmax));
+            }
+            if (scale > 200)
+            {
+                g.setColor(ink3);
+                for (double d = 0.1; d < 0.5; d += 0.1)
+                {
+                    g.draw(new Line2D.Double(u+d, vmin, u+d, vmax));
+                    g.draw(new Line2D.Double(u+0.5+d, vmin, u+0.5+d, vmax));
```

Comment:
Wait, you've hard-coded specific levels of subdivision?

Looks like it works for this use-case, though I'll probably take a different approach if I try to do something similar for the 3d grids...

Reply:
Yes. This design is for this case. In the 3D editor you need an infinitely repeating pattern. I'm pretty sure I had that logic working at some point but it was years ago and that file most likely does not exist any more.... A side track of a side track....